pkg/workloads/cortex/serve/serve.py

```diff
@@ -176,11 +176,15 @@ async def parse_payload(request: Request, call_next):
 
 
 def predict(request: Request):
+    tasks = BackgroundTasks()
     api = local_cache["api"]
     predictor_impl = local_cache["predictor_impl"]
-    args = build_predict_args(request)
+    kwargs = build_predict_kwargs(request)
 
-    prediction = predictor_impl.predict(**args)
+    prediction = predictor_impl.predict(**kwargs)
+    if has_method(predictor_impl, "post_predict"):
+        kwargs = build_post_predict_kwargs(prediction, request)
+        tasks.add_task(predictor_impl.post_predict, **kwargs)
```

Comment:
If there is an exception raised in the post_predict, how does the stack trace look? Is it easy to understand?

Reply:
If an exception is raised in `post_predict`, the stack trace appears to be descriptive enough for the user to understand it originated in `post_predict`. Here's a sample:
```bash
2020-07-30 20:52:31.925351:cortex:pid-19:ERROR:Exception in ASGI application
Traceback (most recent call last):
  File "/opt/conda/envs/env/lib/python3.6/site-packages/uvicorn/protocols/http/httptools_impl.py", line 385, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/fastapi/applications.py", line 151, in __call__
    await super().__call__(scope, receive, send)  # pragma: no cover
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/applications.py", line 102, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/base.py", line 26, in __call__
    await response(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/responses.py", line 197, in __call__
    async for chunk in self.body_iterator:
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/base.py", line 56, in body_stream
    task.result()
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/base.py", line 38, in coro
    await self.app(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/base.py", line 26, in __call__
    await response(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/responses.py", line 197, in __call__
    async for chunk in self.body_iterator:
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/base.py", line 56, in body_stream
    task.result()
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/base.py", line 38, in coro
    await self.app(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/middleware/cors.py", line 76, in __call__
    await self.app(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/routing.py", line 550, in __call__
    await route.handle(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/routing.py", line 44, in app
    await response(scope, receive, send)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/responses.py", line 131, in __call__
    await self.background()
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/background.py", line 35, in __call__
    await task()
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/background.py", line 20, in __call__
    await run_in_threadpool(self.func, *self.args, **self.kwargs)
  File "/opt/conda/envs/env/lib/python3.6/site-packages/starlette/concurrency.py", line 34, in run_in_threadpool
    return await loop.run_in_executor(None, func, *args)
  File "/opt/conda/envs/env/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/mnt/project/predictor.py", line 19, in post_predict
    publish_results()
NameError: name 'publish_results' is not defined
```

What do you think?